### PR TITLE
Pyface Font class and Trait Type

### PR DESCRIPTION
**Pull request by corranwebster**
This is a step in the direction of sorting out #476 

This PR provides:

- a toolkit-independent pure traits `Font` class that roughly corresponds to the CSS font specification
- a simple parser for CSS font specifications
- a `PyfaceFont` trait type that holds such a `Font` instance and auto-converts CSS font descriptions to those fonts
- methods for translating `Font` objects to and from the toolkit font specification objects
- a `FontDialog` and associated classes for bringing up a standard system font dialog to get a `Font` value.
- tests
- basic documentation

### CODE REVIEW

File: pyface/font_dialog.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+""" The implementation of a dialog that allows the user to choose a font.
+"""
+
+from .constant import OK
+from .toolkit import toolkit_object
+
+FontDialog = toolkit_object("font_dialog:FontDialog")
+
+
+def get_font(parent, font):
```
Comment (ievacerny) [inline pyface/font_dialog.py:20]: Shouldn't `parent` be passed to the actual dialog?

Another thing, I might be using this function incorrectly, but simply calling it with `parent=None` and any font, and then clicking on either of the dialog buttons freezes the application with qt. No dialog window is brought up at all if using wx.

And finally, this function isn't tested at all, so might be good to add some tests.

File: pyface/tests/test_font_dialog.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from ..font_dialog import FontDialog
+from ..toolkit import toolkit_object
+
+GuiTestAssistant = toolkit_object("util.gui_test_assistant:GuiTestAssistant")
+no_gui_test_assistant = GuiTestAssistant.__name__ == "Unimplemented"
+
+ModalDialogTester = toolkit_object(
+    "util.modal_dialog_tester:ModalDialogTester"
+)
+no_modal_dialog_tester = ModalDialogTester.__name__ == "Unimplemented"
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/tests/test_font_dialog.py:24]: With a test for get_font being added, this now is used.

File: docs/source/trait_types.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    'italic bold 14pt Helvetica, Arial, sans-serif'
+    '36pt "Comic Sans"'
+
+are valid font descriptions, but "Helvetica bold 12pt" is not because the
+order of elements is wrong.
+
+The |Font| object also has a method :py:meth:`~!Font.to_toolkit` that
+produces a toolkit font specification, which is usually what controls and
+other toolkit-specific code excpect to be given.
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply docs/source/trait_types.rst:82]: I have changed it to "widgets" which should make it clearer.

**Comment by corranwebster** [comment]
Assuming tests pass, this is ready for re-review.

File: pyface/font_dialog.py (hunk 1 above)
Comment (ievacerny) [reply pyface/font_dialog.py:20]: Qt dialog now works fine and wx dialog is brought up. But now it's the wx dialog that freezes for quite some time after a button is clicked (either OK, Cancel). I'm guessing that it's related to the closing of the dialog window.

File: pyface/ui/wx/font_dialog.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        pass
+
+    # ------------------------------------------------------------------------
+    # 'IWindow' interface.
+    # ------------------------------------------------------------------------
+
+    def close(self):
+        font_data = self.control.GetFontData()
+        wx_font = font_data.GetChosenFont()
+        self.font = Font.from_toolkit(wx_font)
```
Comment (ievacerny) [inline pyface/ui/wx/font_dialog.py:46]: Clicking `Cancel` on the dialog raises this error:
```
>>> get_font(None, "Helvetica")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/icernyte/Documents/repos/pyface/pyface/font_dialog.py", line 36, in get_font
    result = dialog.open()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_dialog.py", line 168, in open
    self.close()
  File "/Users/icernyte/Documents/repos/pyface/pyface/ui/wx/font_dialog.py", line 46, in close
    self.font = Font.from_toolkit(wx_font)
  File "/Users/icernyte/Documents/repos/pyface/pyface/font.py", line 387, in from_toolkit
    return cls(**toolkit_font_to_properties(toolkit_font))
  File "/Users/icernyte/Documents/repos/pyface/pyface/ui/wx/font.py", line 162, in toolkit_font_to_properties
    family = wx_family_to_generic_family[toolkit_font.GetFamily()]
wx._core.wxAssertionError: C++ assertion "IsOk()" failed at /Users/robind/projects/bb2/dist-osx-py36/build/ext/wxW
idgets/src/common/fontcmn.cpp(407) in GetFamily(): invalid font
```

File: pyface/font.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+    'semicondensed': 400,
+    'normal': 500,
+    'semi-expanded': 600,
+    'semiexpanded': 600,
+    'expanded': 700,
+    'extra-expanded': 800,
+    'extraexpanded': 800,
+    'ultra-expanded': 900,
+    'ultraexpanded': 900,
+}
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:136]: This looks like a miss on my part- I think somehow I had got into my head that stretches were using the 0 to 1000 values instead of percentages.

File: pyface/font.py [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+        ----------
+        description : str
+            A font description in string form such as
+            'italic bold 14pt Helvetica, Arial, sans-serif' or
+            '36pt "Comic Sans"'
+        """
+        return cls(**parse_font_description(description))
+
+    @classmethod
+    def from_toolkit(cls, toolkit_font):
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:374]: This is because there are going to be situations (like font dialogs, but not only there), where there is some toolkit-level call or object which returns a toolkit font object (either a `QFont` or a `wx.Font`).  We need to be able to convert those to a `Font` object as best we can.  It may not be precise, because different backends have different capabilities, but is should be good enough, and should round-trip within the capabilities of the system.

The fact that it isn't precise is OK because font selection is strongly constrained by what is actually available on on the underlying system, so even if the font selection is not exactly the same, you are likely to get the same typeface when the toolkit asks for things.

[the remark this answers is not in the document]
Comment (kitchoi) [reply pyface/font.py:374]: That makes sense. Thank you! Resolving.

**Comment by corranwebster** [comment]
This `Font` class isn't intended to be subclassed - it's intended to represent a font well enough for (almost) all use cases that a scientific application is likely to need.  It is a leaf class, and deliberately so.  It is not an interface or ABC.

Typography is fantastically fractal and we could imagine any number of things that we might like to support (eg. kerning options, different styles of numerals, etc.) many of which aren't even exposed in Qt, let alone Wx.  I made a deliberate choice to target what CSS can express, as that seems good enough for almost everything we need (it's good enough for the web!), is approximately the intersection of what wxPython (at least in 4.1) and PyQt support, and additionally provides a standard textual representation for the font that is familiar to developers.  The CSS textual representation is important because:

- it is a UX affordance for developers who use the API to be able to specify fonts with a descriptive string rather than a bunch of keywords;
- the intent is to eventually replace the current uses of TraitsUI Font, which accepts a text description in a hacky mini-language, so we would like something similar but better to aid eventual migration
- it can even be exposed to sophisticated users
- for better or worse, it is a standard

So it's quite deliberately limited.  But its likely far more than what we actually _need_ for real scientific computing use-cases.

File: pyface/ui/qt4/font.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    elif toolkit_font.family():
+        family.append(toolkit_font.family())
+    if toolkit_font.defaultFamily():
+        family.append(toolkit_font.defaultFamily())
+    family.append(qt_family_to_generic_family[toolkit_font.styleHint()])
+
+    size = toolkit_font.pointSizeF()
+    style = qt_style_to_style[toolkit_font.style()]
+    weight = map_to_nearest(toolkit_font.weight(), qt_weight_to_weight)
+    stretch = map_to_nearest(toolkit_font.stretch(), qt_stretch_to_stretch)
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/ui/qt4/font.py:177]: I think I want font weights to stay multiples of 100: almost all typefaces outside of specialist typographical fonts are multiples of 100.  Our use case is scientific apps, so I think we're good.

I'll assess the stretch situation separately as there is an underlying problem with the implementation, but even there I think we are constrained by what most fonts in the wild actually support.

These numbers look like the could be arbitrary values, but in practice they are not, and there will never be an actual difference in what appears on a screen or what font you get if you set weight to 501 instead of 500, for example.

File: pyface/ui/qt4/font.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        elif default_family is None:
+            default_family = family
+
+    if families and hasattr(qt_font, 'setFamilies'):
+        # Qt 5.13 and later
+        qt_font.setFamilies(families)
+    elif families:
+        qt_font.setFamily(families[0])
+        # Note: possibily could use substitutions here,
+        # but not sure if global (which would be bad, so we don't)
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/ui/qt4/font.py:133]: This class (along with `QFont` and `wx.Font`) is a font _specification_, not a specific font instance.  What it provides is a way for a developer to state "I'm OK with something that looks like this" and then lets the toolkit and operating system sort out what the best fit is.

When it comes to Enable and Kiva font _rendering_ we will end up caring a lot about this, because for some backends we will have to find the "best fit" font so that we can draw it.  But that's not the point of this `Font` object.

The comment here is that Qt < 5.13 provides a way of specifying "it's OK to use this family rather than that family", but that I think it means messing with global state (ie. what you're specifiying is "it's _always_ OK to use this family rather than that family").  That's more wrong than just taking the first family option.

We definitely _don't_ want to create a `QApplication` as a side-effect.

The `px` problem is a separate issue and should be fixed.

File: pyface/ui/qt4/font.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    'serif': QFont.Serif,
+    'roman': QFont.Serif,
+    'cursive': QFont.Cursive,
+    'script': QFont.Cursive,
+    'sans-serif': QFont.SansSerif,
+    'swiss': QFont.SansSerif,
+    'monospace': QFont.Monospace,
+    'modern': QFont.Monospace,
+    'typewriter': QFont.TypeWriter,
+}
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/ui/qt4/font.py:46]: Because "Swiss" is means "default sans-serif" in ETS' current `Font` traits.

File: pyface/font.py [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+
+    def __init__(self, value=None, **metadata):
+        if isinstance(value, Font):
+            value = value.trait_get(value.editable_traits())
+        if isinstance(value, str):
+            value = parse_font_description(value)
+        if isinstance(value, dict):
+            # freeze the family list and variant set
+            if 'family' in value:
+                value['family'] = tuple(value['family'])
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:468]: We are not going to add more mutable traits to `Font` without a careful redesign, probably prompted in many years time by some important shift in how fonts are represented.

File: pyface/font.py [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+        else:
+            return NotImplemented
+
+
+class PyfaceFont(TraitType):
+    """ A trait that holds a Pyface Font.
+
+    The value can be assigned as a string, in which case it is parsed
+    as a font description and an appropriate font created for it.
+    """
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:455]: I disagree strongly on this.  The `FontDescription` can only be initialized by a string but holds a `Font`, so if you had:
```
class MyClass(HasTraits):
    label_font = FontDescription("Arial")
```
you can't do:
```
foo = MyClass(label_font="Comic Sans")
```
unless you also modify the validator to also do parsing, but then you get the weird cases of
```
foo.label_font = foo.label_font
```
raising a `TraitError` unless you _also_ accept `Font` objects as valid, and then why shouldn't you be able to say
```
class MyClass(HasTraits):
    label_font = FontDescription(Font(family=["Arial"]))
```
when you can say
```
foo.label_font = Font(family=["Arial"])
```
And doing a sub-class
```
class MySubClass(MyClass):
    label_font = ????  # what is valid here?
```
is even weirder.

Basically `PyfaceFont` is what you get when you want to support text-to-Font conversion (and we do!).  There is still `Instance(Font)` if you _don't_ want string text conversion.

File: pyface/font.py [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+
+    def __repr__(self):
+        traits = self.trait_get(self.editable_traits())
+        trait_args = ', '.join(
+            "{}={!r}".format(name, value)
+            for name, value in traits.items()
+        )
+        return "{}({})".format(self.__class__.__name__, trait_args)
+
+    def __eq__(self, other):
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:436]: Yep - deliberate.  The class is mutable, therefore is not hashable.

And I think that having `__eq__` implemented is desirable in production code.  In particular, if something round-trips to toolkit code or a string and does not change (eg.a future text-style TraitsUI `FontEditor`) then nothing has really changed.

File: pyface/font.py [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+            value = parse_font_description(value)
+        if isinstance(value, dict):
+            # freeze the family list and variant set
+            if 'family' in value:
+                value['family'] = tuple(value['family'])
+            if 'variants' in value:
+                value['variants'] = frozenset(value['variants'])
+            default_value = (Font, (), value.copy())
+        elif value is None:
+            default_value = (Font, (), {})
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:473]: See other comments.  `Font` isn't designed to be and shouldn't be subclassed.

File: pyface/font.py [hunk 14]
```diff
@@ -0,0 +1,478 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:1]: No, but I might be able to be convinced that CSS-style font description parsing should be in a helper module.

File: pyface/font.py [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+
+    #: The size of the font.
+    size = FontSize()
+
+    #: The font variants.
+    variants = FontVariants()
+
+    @classmethod
+    def from_description(cls, description):
+        """ An extremely lax 'parser' for CSS-style font descriptions.
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:362]: Maybe this should be renamed "from_css" or something, but this is not a class that is intended to be subclassed.

File: pyface/font.py [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+    'small': 10.0,
+    #'medium': 12.0,
+    'large': 14.0,
+    'x-large': 18.0,
+    'xx-large': 20.0,
+}
+
+styles = ['normal', 'italic', 'oblique']
+
+variants = ['small-caps', 'underline', 'strikethrough', 'overline']
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:150]: I'm personally in two minds about all caps for container module "constants".  If they are intended to be mutated (eg. a registry) then I don't like all-caps.  In this case it's probably fine (in fact, I should possibly make them tuples rather than lists.

File: pyface/font.py [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+
+#: A trait for font families.
+FontFamily = List(Str, ['default'])
+
+#: A trait for font weights.
+FontWeight = Map(weights, default_value='normal')
+
+#: A trait for font stretch values.
+FontStretch = Map(stretches, default_value='normal')
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/font.py:159]: I think I _want_ this as a design choice.  I certainly do for `FontWeight`.

File: pyface/ui/wx/font_dialog.py (hunk 4 above)
Comment (corranwebster) [reply pyface/ui/wx/font_dialog.py:46]: I'll have a play with this - I think it might require exposing (at least at the private level) methods for getting and setting the font, but I'm OK with that as they can then be used to push common code into the mixin.

File: pyface/ui/qt4/font.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply pyface/ui/qt4/font.py:133]: Agreed that we don't want to have the side-effect of creating `QApplication`.  In fact, that wasn't what I was suggesting to do. I think `QApplication.instance()` will return `None` if the `QApplication` has not been created, and in that case, definitely don't create a QApplication. But if the returned value is not None, then we can loop over the family instead of picking the first one which may not work (whereas the second/third/... might). That is a nice-to-have I think. At the very least, I think this branch could log a warning saying we are picking off the first one and ignoring the rest (if there are any).

**Comment by corranwebster** [comment]
Changes I am planning to make based on all of this:

- [ ] a look at font stretch, because I think I messed up the shadowed values there; and to check if it makes sense to allow arbitrary values
- [ ] extend parsing a bit so actual all CSS font descriptors are actually parsed correctly, and then complain if we can't handle something (eg. relative font sizes).
- [ ] add some utility methods for accessing the font on a font dialog and refactor to use them both in implementation and tests
- [ ] add some testing for backends for `from_toolkit` and `to_toolkit`

File: pyface/font.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply pyface/font.py:1]: OK. Moving the helper functions around can happen later I guess, when this module needs to grow and be reorganized.

File: pyface/ui/qt4/font.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply pyface/ui/qt4/font.py:133]: Just to be more specific about what I was thinking about. This does not have to happen here, but an idea for consideration:
```
def find_family(families): 
    if len(families) == 1: 
        return families[0] 
    app = QtGui.QApplication.instance() 
    if app is None: 
        logger.warning("using the first one only") 
        return families[0] 
    for family in families: 
        if QtGui.QFontInfo(QtGui.QFont(family)).family() == family: 
            return family 
    logger.warning("using the first one only after trying all.") 
    return families[0]
```
e.g.
```
# QApplication has not been created
>>> find_family(["Not Found", "Arial"])
using the first one only
'Not Found'
>>> QtGui.QApplication([])
>>> find_family(["Arial"])
"Arial" 
```
For consideration only as I have to agree that this is a bit magical, just to support something missing in Qt < 5.13.

File: pyface/font.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kitchoi) [reply pyface/font.py:455]: My bad for omitting the `validate` in my `FontDescription`. My apologies. I was mainly thinking about how we could make `__init__` simpler and reuse what is already available from `Instance`.  So I omitted the `validate` due to laziness. For a trait type responsible for text-to-font conversion, I would still expect the text-to-font conversion to be included in the `validate` as well.

I see that if one wants to support text-to-font conversion in defining a trait, the trait type needs to support what `Instance(Font)` also supports, and the validation will have to be the combination of the two.

I have to admit I am not fond of the text-to-font conversion support. I find defining `Font(family=["Arial"], variants={"small-caps"}, weight="bold")` safe, readable, intuitive and explicit. I can read the trait class definition and then I know what to write, I don't have to learn a relaxed CSS-like language. In other words, the convenience of typing less does not give me enough confidence and I would still prefer defining `Font` traits explicitly. Therefore the conversion layer strikes me as a rather costly thing to support.

I am probably alone in that, so I will continue to review accepting this will be a popular feature.
